crates/compiler/mono/src/borrow.rs

```diff
@@ -1,119 +1,401 @@
+        procs: procs
+            .iter()
+            .map(|(_key, proc)| {
+                let key = (proc.name.name(), proc.proc_layout(arena));
+                let signature = BorrowSignature::from_layouts(interner, key.1.arguments.iter());
+                (key, signature)
+            })
+            .collect(),
+    };
+
+    // for every proc (by index) a collection of its join points
+    let mut join_points: Vec<_> = std::iter::repeat_with(MutMap::default)
+        .take(procs.len())
+        .collect_in(arena);
+
+    // next we first partition the functions into strongly connected components, then do a
+    // topological sort on these components, finally run the fix-point borrow analysis on each
+    // component (in top-sorted order, from primitives (std-lib) to main)
+
+    let matrix = construct_reference_matrix(arena, procs);
+    let sccs = matrix.strongly_connected_components_all();
+
+    let mut join_point_stack = Vec::new_in(arena);
+    let mut proc_join_points = MutMap::default();
+
+    for (group, _) in sccs.groups() {
+        // This is a fixed-point analysis
+        //
+        // all functions initially own all their parameters
+        // through a series of checks and heuristics, some arguments are set to borrowed
+        // when that doesn't lead to conflicts the change is kept, otherwise it may be reverted
+        //
+        // when the signatures no longer change, the analysis stops and returns the signatures
+
+        loop {
+            let mut modified = false;
+
+            for index in group.iter_ones() {
+                let (_, proc) = procs.iter().nth(index).unwrap();
+                let key = (proc.name.name(), proc.proc_layout(arena));
+
+                if proc.args.is_empty() {
+                    continue;
+                }
+
+                std::mem::swap(&mut proc_join_points, &mut join_points[index]);
+
+                let mut state = State {
+                    args: proc.args,
+                    borrow_signature: *borrow_signatures.procs.get(&key).unwrap(),
+                    join_point_stack,
+                    join_points: proc_join_points,
+                    modified: false,
+                };
+
+                state.inspect_stmt(interner, &mut borrow_signatures, &proc.body);
+
+                // did any proc signature get modified?
+                //
+                // NOTE: this does not directly include updates to join point signatures. The
+                // assumption is that a relevant change in join point signature is immediately
+                // (i.e. no fixpoint is required) reflected in the proc signature.
+                assert_eq!(
+                    state.modified,
+                    borrow_signatures
+                        .procs
+                        .insert(key, state.borrow_signature)
+                        .unwrap()
+                        != state.borrow_signature
+                );
```

Comment:
@bhansconnect this is the assert in question. If you

- replace this with a `debug_assert_eq` (or remove it entirely I think)
- build the compiler in release mode
- run the repl in release mode
- run some basic code like `Str.concat "foo" "bar"`

this runs into some infinite loop, my guess is in the borrow inference where it iterates over the result of topological sort.

this was on ubuntu but it caused issues on other OSes as well.

Reply:
I say, lets just submit it with a TODO to debug more later.

Reply:
ok comment added. This should be good to go then (for now)